R/run.R

```diff
@@ -22,33 +27,55 @@
 #'
 #' # run basic example within learnr
 #' \dontrun{run_tutorial("hello", "learnr")}
-run_tutorial <- function(name = NULL, package = NULL, shiny_args = NULL) {
+run_tutorial <- function(
+  name = NULL,
+  package = NULL,
+  shiny_args = NULL,
+  clean = FALSE
+) {
+  checkmate::assert_character(name, any.missing = FALSE, max.len = 1, null.ok = TRUE)
+  checkmate::assert_character(package, any.missing = FALSE, max.len = 1, null.ok = TRUE)
 
-  if (is.null(package) && !is.null(name)) {
-    stop.("`package` must be provided if `name` is provided.")
+  # is `name` a valid and existing directory for `rmarkdown::run()`?
+  name <- validate_tutorial_path_is_dir(name)
+  name_is_tutorial_path <- is.null(package) && name$valid
+  name <- name$value
+
+  if (!name_is_tutorial_path && !is.null(name) && is.null(package)) {
+    stop.("`package` must be provided if `name` is the name of a packaged tutorial. Otherwise, `name` must be a directory.")
   }
 
   # works for package = NULL and if package is provided
-  tutorials <- available_tutorials(package = package)
-  if (is.null(name)) {
+  if (!name_is_tutorial_path && is.null(name)) {
+    tutorials <- available_tutorials(package = package)
     message(format(tutorials))
     return(invisible(tutorials))
   }
 
-  # get path to tutorial
-  tutorial_path <- get_tutorial_path(name, package)
+  # get path to tutorial directory
+  tutorial_path <-
+    if (name_is_tutorial_path) {
+      normalizePath(name)
+    } else {
+      get_tutorial_path(name, package)
+    }
 
   # check for necessary tutorial package dependencies
   install_tutorial_dependencies(tutorial_path)
 
   # provide launch_browser if it's not specified in the shiny_args
-  if (is.null(shiny_args))
+  if (is.null(shiny_args)) {
     shiny_args <- list()
+  }
   if (is.null(shiny_args$launch.browser)) {
-    shiny_args$launch.browser <- (
-      interactive() ||
-        identical(Sys.getenv("LEARNR_INTERACTIVE", "0"), "1")
-    )
+    is_interactive <- interactive() ||
```

Comment:
Do we import `rlang`? If so....
```suggestion
    is_interactive <- rlang::is_interactive() ||
```

Reply:
yeah I let this slide, along with a few other places where I would reach for `rlang` because I was thinking it would be better to do an `rlang` upgrade as a concerted effort rather than piecemeal. In this case though I think you're right that it's worth the benefit.

Reply:
It could be left as a TODO for later. (The change is independent of the PR goals)